apiserver/remotefirewaller/ingressaddresswatcher.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/juju/juju/network"
+	"github.com/juju/juju/state/watcher"
+	"github.com/juju/juju/worker/catacomb"
+)
+
+// IngressAddressWatcher reports changes to addresses
+// for local units in a given relation.
+// Each event contains the entire set of addresses which
+// are required for ingress for the relation.
+type IngressAddressWatcher struct {
+	catacomb catacomb.Catacomb
+
+	backend State
+	appName string
+	rel     Relation
+
+	out chan []string
+
+	// A map of known unit addresses, keyed on unit name.
+	known map[string]string
+}
+
+// NewIngressAddressWatcher creates an IngressAddressWatcher.
+func NewIngressAddressWatcher(backend State, rel Relation, appName string) (*IngressAddressWatcher, error) {
+	w := &IngressAddressWatcher{
+		backend: backend,
+		appName: appName,
+		rel:     rel,
+		known:   make(map[string]string),
+		out:     make(chan []string),
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	})
+	return w, err
+}
+
+func (w *IngressAddressWatcher) loop() error {
+	ruw, err := w.rel.WatchUnits(w.appName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(ruw); err != nil {
+		return errors.Trace(err)
+	}
+
+	var (
+		sentInitial bool
+		out         chan<- []string
+	)
+	// addresses holds the current set of known addresses.
+	addresses := make(set.Strings)
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case c, ok := <-ruw.Changes():
+			if !ok {
+				return watcher.EnsureErr(ruw)
+			}
+			// A unit has entered or left scope.
+			// Get the new set of addresses resulting from that
+			// change, and if different to what we know, send the change.
+			merged, err := w.mergeAddresses(addresses, c)
+			if err != nil {
+				return err
+			}
+			changed := !merged.Difference(addresses).IsEmpty() || !addresses.Difference(merged).IsEmpty()
+			if !sentInitial || changed {
+				addresses = merged
+				out = w.out
+			} else {
+				out = nil
+			}
+		case out <- formatAsCIDR(addresses.Values()):
+			sentInitial = true
+			out = nil
+		}
+	}
+}
+
+func formatAsCIDR(addresses []string) []string {
+	result := make([]string, len(addresses))
+	for i, a := range addresses {
+		result[i] = a + "/32"
+	}
+	return result
+}
+
+func (w *IngressAddressWatcher) Changes() <-chan []string {
+	return w.out
+}
+
+func (w *IngressAddressWatcher) mergeAddresses(addresses set.Strings, c params.RelationUnitsChange) (set.Strings, error) {
+	result := set.NewStrings(addresses.Values()...)
+	for name := range c.Changed {
+
+		u, err := w.backend.Unit(name)
+		if errors.IsNotFound(err) {
+			logger.Warningf("unit %s is not found, can't get address", name)
+			continue
+		}
+		if err != nil {
+			return result, err
+		}
+
+		// TODO - start watcher to pick up machine address changes
+		// We need to know whether to look at the public or cloud local address.
+		// For now, we'll use the public address and later if needed use a watcher
```

Comment:
that's probably fine in general, but we still need to watch addresses in case (a) the machine doesn't have an address yet, or (b) the public address changes